app/src/main/java/com/example/woloxapp/ui/login/LoginFragment.kt

```diff
@@ -62,19 +63,42 @@ class LoginFragment : Fragment() {
                 }
             }
             loginViewModel.validEmail.observe(viewLifecycleOwner) {
-                if (it!!) loginViewModel.login(User(userName.text.toString(), password.text.toString()))
-                else binding.userName.error = getString(R.string.invalid_username)
+                if (it!!) fetchData(User(userName.text.toString(), password.text.toString()))
             }
-            loginViewModel.userIsLogged.observe(viewLifecycleOwner) {
-                if (it) findNavController().navigate(R.id.go_to_home)
-                else showToast(R.string.login_failure)
+        }
+    }
+    private fun fetchData(user: User) {
+        loginViewModel.login(user)
+        binding.progressBar.visibility = View.VISIBLE
+        loginViewModel.response.observe(viewLifecycleOwner) {
+                response ->
+            when (response) {
+                is NetworkResult.Success -> {
+                    this.findNavController()?.navigate(R.id.HomePageFragment)
+                }
+                is NetworkResult.Error -> {
+                    response.message?.let { showToast(it) }
+                    binding.progressBar.visibility = View.INVISIBLE
+                }
+                is NetworkResult.Loading -> {
```

Comment:
The network state is a bit weird for this scenario, if the api response is Loading you are stuck here with an infinite progressBar, also i don't think this api can return "loading".